ext/random/random.c

```diff
@@ -680,35 +680,82 @@ uint64_t php_random_generate_fallback_seed(void)
 		gettimeofday(&tv, NULL);
 		fallback_seed_add(&c, &tv, sizeof(tv));
 		/* Previous state. */
-		fallback_seed_add(&c, RANDOM_G(fallback_seed), 20);
+		fallback_seed_add(&c, state->seed, 20);
 	}
-	PHP_SHA1Final(RANDOM_G(fallback_seed), &c);
-	RANDOM_G(fallback_seed_initialized) = true;
+	PHP_SHA1Final(state->seed, &c);
+	state->initialized = true;
 
 	uint64_t result = 0;
 
 	for (int i = 0; i < sizeof(result); i++) {
-		result = result | (((uint64_t)RANDOM_G(fallback_seed)[i]) << (i * 8));
+		result = result | (((uint64_t)state->seed[i]) << (i * 8));
 	}
 
 	return result;
 }
 
-/* {{{ PHP_GINIT_FUNCTION */
-static PHP_GINIT_FUNCTION(random)
+uint64_t php_random_generate_fallback_seed(void)
 {
-	random_globals->random_fd = -1;
-	random_globals->fallback_seed_initialized = false;
+	return php_random_generate_fallback_seed_ex(&RANDOM_G(fallback_seed_state));
 }
-/* }}} */
 
-/* {{{ PHP_GSHUTDOWN_FUNCTION */
-static PHP_GSHUTDOWN_FUNCTION(random)
+static zend_result php_general_random_bytes_for_zend_initialize(php_random_state_for_zend *state)
+{
+	uint64_t t[4];
+
+	do {
+		char errstr[128];
+		if (php_random_bytes_ex(&t, sizeof(t), errstr, sizeof(errstr)) == FAILURE) {
+#if ZEND_DEBUG
+			fprintf(stderr, "php_random_bytes_ex: Failed to generate a random seed: %s\n", errstr);
+#endif
+			return FAILURE;
+		}
+	} while (UNEXPECTED(t[0] == 0 && t[1] == 0 && t[2] == 0 && t[3] == 0));
+
+	php_random_xoshiro256starstar_seed256(&state->xoshiro256starstar_state, t[0], t[1], t[2], t[3]);
+
+	return SUCCESS;
+}
+
+static void php_general_random_bytes_for_zend_initialize_fallback(php_random_state_for_zend *state)
+{
+	uint64_t t;
+	php_random_fallback_seed_state fallback_state;
+
+	do {
+		t = php_random_generate_fallback_seed_ex(&fallback_state);
+	} while (UNEXPECTED(t == 0));
+
+	php_random_xoshiro256starstar_seed64(&state->xoshiro256starstar_state, t);
+}
+
+PHPAPI zend_result php_general_random_bytes_for_zend(zend_utility_general_random_state *opaque_state, void *bytes, size_t size)
 {
-	if (random_globals->random_fd >= 0) {
-		close(random_globals->random_fd);
-		random_globals->random_fd = -1;
+	php_random_state_for_zend *state = (php_random_state_for_zend*) opaque_state;
+
+	if (!state->initialized) {
+		if (php_general_random_bytes_for_zend_initialize(state) == FAILURE) {
+			php_general_random_bytes_for_zend_initialize_fallback(state);
+		}
+		state->initialized = true;
+	}
+
+	while (size > 0) {
+		php_random_result result = php_random_algo_xoshiro256starstar.generate(&state->xoshiro256starstar_state);
+		size_t chunk_size = MIN(size, sizeof(result.size));
+		memcpy(bytes, &result.result, chunk_size);
+		size -= chunk_size;
+		bytes = (char*)bytes + chunk_size;
```

Comment:
The current code is broken:

- `sizeof(result.size)` is always `8`.
- But even if this would be fixed to use `result.size`, it would be broken on big endian architectures if `result.size < 8`, because then you're copying the undefined upper bytes of `result.result`.

The following would match the implementation of `Randomizer::getBytes()`:

```
PHPAPI zend_result php_general_random_bytes_for_zend(zend_utility_general_random_state *opaque_state, char *bytes, size_t length)
{
	php_random_state_for_zend *state = (php_random_state_for_zend*) opaque_state;

	if (!state->initialized) {
		php_general_random_bytes_for_zend_initialize(state);
		state->initialized = true;
	}

	size_t total_size = 0;
	while (total_size < length) {
		php_random_result result = php_random_algo_xoshiro256starstar.generate(&state->xoshiro256starstar_state);
		for (size_t i = 0; i < result.size; i++) {
			bytes[total_size++] = (result.result >> (i * 8)) & 0xff;
			if (total_size >= length) {
				break;
			}
		}
	}

	return SUCCESS;
}
```

Alternatively you can rely on the fact that xoshiro256** always returns 8 bytes and `memcpy` [1] away, but then you shouldn't use `result.size`.

[1] zend_mempcpy would make the code a little nicer.